### PR TITLE
Data graph walking, searching, and manipulation functions from clojure-navigation library

### PR DESCRIPTION
**Pull request by coconutpalm**
Cleaned up clojure-navigation code; haven't written tests yet.

### CODE REVIEW

**Comment by coconutpalm** [comment]
Since this is copied code, I didn't add tests, but did add a tech debt story to write tests: https://www.pivotaltracker.com/n/projects/1354106/stories/128825149

**Comment by recursive-automata** [comment]
👍 , but unit tests recommended ;)

File: src/clj_foundation/conversions.clj [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  * (convert Boolean/TYPE \"true\")
+  * (convert Map vararg-parameter-kvs)"
+  (fn [dest-class src-instance] [dest-class (class src-instance)]))
+
+
+(defmethod convert [java.io.File String] [_ str]
+   (java.io.File. str))
+
+
+(defmethod convert [Boolean/TYPE String] [_ str]
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/conversions.clj]: Not sure how you're using this yet, but I would add support for int values 0, 1, and string values "0" and "1".

File: src/clj_foundation/conversions.clj [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  (contains? #{"on" "yes" "true"} (.toLowerCase str)))
+
+
+(defmethod convert [Map Vector] [_ v]
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/conversions.clj]: Doesn't "zipmap" already do this? https://clojuredocs.org/clojure.core/zipmap

I see it's a little different input and returns HashMap, but same functionality: (class (zipmap [:a :b] [1 2]))

File: src/clj_foundation/grep.clj [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            [clj-foundation.tree-visit :as v]
+            [clj-foundation.patterns :refer [nothing]])
+  (:import [java.util Map Map$Entry]))
+
+
+(defn- result-coll?
+  "Don't return Map$Entrys as results; other collection types are supported."
+  [node]
+  (and (not (instance? Map$Entry node))
+       (or (map? node) (vector? node) (list? node) (set? node) (seq? node))))
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/grep.clj]: Maybe define after traversable-coll? and use for this line of evaluation?

File: src/clj_foundation/grep.clj [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+
+(defn- traversable-coll?
+  "Is this a kind of collection we can traverse?"
+  [node]
+  (or (map? node) (vector? node) (list? node) (set? node) (seq? node)))
+
+
+(defn root-node?
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/grep.clj]: I might consider adding metadata for functions with return type of $zipObject throughout the library.

File: src/clj_foundation/grep.clj [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+
+(defn- result-coll?
+  "Don't return Map$Entrys as results; other collection types are supported."
+  [node]
+  (and (not (instance? Map$Entry node))
+       (or (map? node) (vector? node) (list? node) (set? node) (seq? node))))
+
+
+(defn- traversable-coll?
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/grep.clj]: These generic type function I would make public.

File: src/clj_foundation/conversions.clj (hunk 2 above)
Comment (coconutpalm) [reply src/clj_foundation/conversions.clj]: No: this is for the case where you want something like
[:a 1 :b 2] --> {:a 1 :b 2}

File: src/clj_foundation/patterns.clj [hunk 6]
```diff
@@ -41,6 +41,15 @@
        (apply str)))
 
 
+(defn arity
+ "Returns the maximum parameter count of each invoke method found by refletion
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/patterns.clj]: typo: refletion --> reflection

File: src/clj_foundation/grep.clj [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  [state breadcrumb loc]
+  {:state (conj state [breadcrumb loc])})
+
+
+(defn- grep-match?
+  "Is the current node a match?  If so, capture it in state using found-match-fn."
+  [pattern node state loc breadcrumb found-match-fn]
+  (if-not (or (v/traversable-coll? node) (nil? node))
+    (cond
+      (and (instance? java.util.regex.Pattern pattern) (re-matches pattern (.toString node))) (found-match-fn state breadcrumb loc)
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/grep.clj]: I might default to "re-find" or add an option to use re-find, as it is more "greppy"

File: src/clj_foundation/grep.clj [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      (and (string? pattern) (.contains (.toString node) pattern))                            (found-match-fn state breadcrumb loc)
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/grep.clj]: Might lowercase search and searched strings, or add option for case insensitive string matching.

File: src/clj_foundation/transducers.clj [hunk 9]
```diff
@@ -0,0 +1,89 @@
+(ns clj-foundation.transducers
+  "Utilities helping to make transducer functions out of regular functions.
+  Transducers can be piplined like thread-first or thread-list, but using regular
```
Comment (levi-andrew-dixon) [inline src/clj_foundation/transducers.clj]: typo:piplined

**Comment by levi-andrew-dixon** [comment]
👍

**Pull request merged**